### PR TITLE
Add app tag apis

### PR DESCRIPTION
**Pull request by duanxianbo**
[(task:19106)](https://staging.teamyapp.com/teams/107/tasks/19106)

### CODE REVIEW

File: core/api/gql/schema.graphql [hunk 1]
```diff
@@ -82,6 +82,11 @@ type Mutation {
 	# App
 	createApp(teamId: ID!, name: String!): App!
 	deleteApp(appId: ID!): App!
+	
+	# App Tag
+	addAppTag(appId: ID!, tagName: String!): Tag!
```
Comment (naturetime10) [inline core/api/gql/schema.graphql]:
```suggestion
	addTagToApp(appId: ID!, tag: String!): Tag!
```

File: core/api/gql/schema.graphql [hunk 2]
```diff
@@ -82,6 +82,11 @@ type Mutation {
...
+	removeAppTag(appId: ID!, tagId: ID!): Tag!
```
Comment (naturetime10) [inline core/api/gql/schema.graphql]:
```suggestion
	removeTagFromApp(appId: ID!, tagId: ID!): Tag!
```

File: core/api/gql/schema.graphql [hunk 3]
```diff
@@ -329,6 +334,12 @@ type App {
 	teamRollouts: [Rollout!]!
 	managedByTeam: Team!
 	latestVersionForTeam(teamId: ID!): AppVersion
+	tags: [Tag!]!
+}
+
+type Tag {
+	id: ID!
+	name: String!
```
Comment (naturetime10) [inline core/api/gql/schema.graphql]:
```suggestion
	tag: String!
```

File: core/api/gql/tag.go [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+func (t Tag) ID(ctx context.Context) graphql.ID {
+	return toGraphQLID(t.tag.ID)
+}
+
+func (t Tag) Name(ctx context.Context) string {
+	return t.tag.Name
+}
+
+func (m Mutation) AddAppTag(
```
Comment (naturetime10) [inline core/api/gql/tag.go]:
```suggestion
func (m Mutation) AddTagToApp(
```

File: core/api/gql/tag.go [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+	if internalErr != nil {
+		internalErr := errs.NewError(
+			errs.InvalidArgument,
+			internalErr.Error(),
+		)
+		m.deps.logger.ErrorWithContext(ctx, internalErr)
+		return Tag{}, errs.ToResolverErr(internalErr)
+	}
+
+	tag, err := m.deps.appService.AddAppTag(ctx, appID, args.TagName)
```
Comment (naturetime10) [inline core/api/gql/tag.go]:
```suggestion
	tag, err := m.deps.appService.AddTagToApp(ctx, appID, args.TagName)
```

File: core/api/gql/tag.go [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+	if err != nil {
+		m.deps.logger.ErrorWithContext(ctx, err)
+		return Tag{}, errs.ToResolverErr(err)
+	}
+
+	return newTag(m.deps, tag), nil
+}
+
+func (m Mutation) RemoveAppTag(
```
Comment (naturetime10) [inline core/api/gql/tag.go]:
```suggestion
func (m Mutation) RemoveTagFromApp(
```

File: core/api/gql/tag.go [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+	if internalErr != nil {
+		internalErr := errs.NewError(
+			errs.InvalidArgument,
+			internalErr.Error(),
+		)
+		m.deps.logger.ErrorWithContext(ctx, internalErr)
+		return Tag{}, errs.ToResolverErr(internalErr)
+	}
+
+	tag, err := m.deps.appService.RemoveAppTag(ctx, appID, tagID)
```
Comment (naturetime10) [inline core/api/gql/tag.go]:
```suggestion
	tag, err := m.deps.appService.RemoveTagFromApp(ctx, appID, tagID)
```

File: core/dao/app_tag_relation.go [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+import (
+	"context"
+
+	"github.com/teamyapp/cloud/libs/errs"
+	"github.com/teamyapp/cloud/libs/transaction"
+	"github.com/teamyapp/teamy-backend/core/entity"
+)
+
+type AppTagRelation interface {
+	FindAppTagByAppIDAndTagIDRelationWithTx(ct context.Context, tx *transaction.Transaction, appID uint64, tagID uint64) (entity.AppTagRelation, *errs.Error)
```
Comment (naturetime10) [inline core/dao/app_tag_relation.go]:
```suggestion
	FindAppTagByAppIDAndTagIDRelationWithTx(
	    ct context.Context, 
	    tx *transaction.Transaction, 
	    appID uint64, 
	    tagID uint64,
    ) (entity.AppTagRelation, *errs.Error)
```

File: core/dao/app_tag_relation.go [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+	FindTagIDsByAppIDWithTx(ct context.Context, tx *transaction.Transaction, appID uint64) ([]uint64, *errs.Error)
```
Comment (naturetime10) [inline core/dao/app_tag_relation.go]:
```suggestion
	FindTagIDsByAppIDWithTx(
	    ct context.Context, 
	    tx *transaction.Transaction, 
	    appID uint64,
	) ([]uint64, *errs.Error)
```

File: core/dao/app_tag_relation.go [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+	CreateAppTagRelation(ct context.Context, tx *transaction.Transaction, appTagRelation entity.AppTagRelation) *errs.Error
+	DeleteAppTagRelationByAppIDAndTagID(ct context.Context, tx *transaction.Transaction, appID uint64, tagID uint64) (entity.AppTagRelation, *errs.Error)
```
Comment (naturetime10) [inline core/dao/app_tag_relation.go]:
```suggestion
	CreateAppTagRelation(
	    ct context.Context, 
	    tx *transaction.Transaction, 
	    appTagRelation entity.AppTagRelation) *errs.Error
	DeleteAppTagRelationByAppIDAndTagID(
	    ct context.Context, 
	    tx *transaction.Transaction, 
	    appID uint64, 
	    tagID uint64,
	) (entity.AppTagRelation, *errs.Error)
```

File: core/dao/sqldb/app_tag_relation.go [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+	ct context.Context,
+	tx *transaction.Transaction,
+	appID uint64,
+	tagID uint64,
+) (entity.AppTagRelation, *errs.Error) {
+	appTagRelation := entity.AppTagRelation{}
+	err := tx.SQLTx().QueryRow(`
+		DELETE FROM app_tag_relation
+		WHERE app_id = $1 AND tag_id = $2
+		RETURNING app_id, tag_id;`,
```
Comment (naturetime10) [inline core/dao/sqldb/app_tag_relation.go]:
```suggestion
;`,
```

Comment (naturetime10) [reply core/dao/sqldb/app_tag_relation.go]: We already have `appID` and `tagID`. Don't need to return them.

File: core/dao/sqldb/tag.go [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+		id,
+		name
+	FROM tag
+	WHERE id IN (%s);`,
+		idsString)
+	rows, err := tx.SQLTx().QueryContext(ct, query)
+	if err != nil {
+		return nil, errs.NewError(errs.Unknown, err.Error())
+	}
+	defer rows.Close()
```
Comment (naturetime10) [inline core/dao/sqldb/tag.go]:
```suggestion
	}
	
	defer rows.Close()
```

File: core/service/app.go [hunk 13]
```diff
@@ -80,8 +82,11 @@ type CreateAppSecretInput struct {
 }
 
 type UpdateAppSecretInput struct {
-	AppSecretID uint64
-	Name        string
+	Name string
+}
+
+type UpdateAppInput struct {
+	Tags []string
```
Comment (naturetime10) [inline core/service/app.go:89]: Do we have enough input to update tags for an app?

File: core/service/app.go [hunk 14]
```diff
@@ -332,6 +337,124 @@ func (a App) CreateApp(ct context.Context, name string, teamID uint64) (entity.A
...
+			return err
+		}
+
+		return nil
+	})
+
+	return tags, err
+}
+
+func (a App) AddAppTag(ct context.Context, appID uint64, tagName string) (entity.Tag, *errs.Error) {
```
Comment (naturetime10) [inline core/service/app.go]:
```suggestion
func (a App) AddTagToApp(ct context.Context, appID uint64, tagName string) (entity.Tag, *errs.Error) {
```

File: core/service/app.go [hunk 15]
```diff
@@ -332,6 +337,124 @@ func (a App) CreateApp(ct context.Context, name string, teamID uint64) (entity.A
...
+			TagID: tag.ID,
+		}
+
+		return a.appTagRelationDao.CreateAppTagRelation(ct, tx, appTagRelation)
+	})
+
+	return tag, err
+}
+
+func (a App) RemoveAppTag(ct context.Context, appID uint64, tagID uint64) (entity.Tag, *errs.Error) {
```
Comment (naturetime10) [inline core/service/app.go]:
```suggestion
func (a App) RemoveTagFromApp(ct context.Context, appID uint64, tagID uint64) (entity.Tag, *errs.Error) {
```

File: migrations/20240120221344_creat_app_tag_table.sql [hunk 16]
```diff
@@ -0,0 +1,16 @@
...
+CREATE TABLE app_tag_relation (
+    app_id BIGINT NOT NULL REFERENCES app (id) on DELETE CASCADE,
+    tag_id BIGINT NOT NULL REFERENCES tag (id) on DELETE CASCADE,
+    CONSTRAINT pk_app_tag_relation PRIMARY KEY (app_id, tag_id)
+);
+
+-- +migrate Down
+DROP TABLE app_tag_relation;
+DROP TABLE tag;
+
```
Comment (naturetime10) [inline migrations/20240120221344_creat_app_tag_table.sql]:
```suggestion
```

**Review by naturetime10 — DISMISSED** [review]

File: core/service/app.go (hunk 13 above)
Comment (duanxianbo) [reply core/service/app.go:89]: this is deprecated, I will remove this

**Review by naturetime10 — APPROVED** [review]

**Pull request merged**